Bug 1142855 - Soft Key Implementation

pgmanutd opened the pull request:
Added softkey utility library. This is just a prototype.
Please let us know if we are on right track.
There are some TODO's. Please clarify those.

HTML Markup:-
&lt;div role="softkey"&gt;
&lt;button type="button" accesskey="a"&gt;Back&lt;/button&gt;
&lt;button type="button" accesskey="s"&gt;OK&lt;/button&gt;
&lt;button type="button" accesskey="d" contextmenu="menu1"&gt;Options&lt;/button&gt;
&lt;menu id="menu1" header="Header Text"&gt;
&lt;menuitem data-menuid="option1" label="Option1"
icon="img/icons/icon48x48.png"&gt;&lt;/menuitem&gt;
&lt;menuitem data-menuid="option2" label="Option2"
icon="img/icons/icon48x48.png"&gt;&lt;/menuitem&gt;
&lt;menuitem data-menuid="option3" label="Option3"
icon="img/icons/icon48x48.png"&gt;&lt;/menuitem&gt;
&lt;menuitem data-menuid="option4" label="Option4"
icon="img/icons/icon48x48.png"&gt;&lt;/menuitem&gt;
&lt;/menu&gt;
&lt;/div&gt;

Commit:
Bug 1142855 - Soft Key Implementation

Added softkey utility library. This is just a prototype.
Please let us know if we are on right track.
There are some TODO's. Please clarify those.

HTML Markup:-
<div role="softkey">
<button type="button" accesskey="a">Back</button>
<button type="button" accesskey="s">OK</button>
<button type="button" accesskey="d" contextmenu="menu1">Options</button>
<menu id="menu1" header="Header Text">
<menuitem data-menuid="option1" label="Option1"
icon="img/icons/icon48x48.png"></menuitem>
<menuitem data-menuid="option2" label="Option2"
icon="img/icons/icon48x48.png"></menuitem>
<menuitem data-menuid="option3" label="Option3"
icon="img/icons/icon48x48.png"></menuitem>
<menuitem data-menuid="option4" label="Option4"
icon="img/icons/icon48x48.png"></menuitem>
</menu>
</div>

AlekoNN commented:
@harman-red-square/harman-reviewers

davidflanagan commented on shared/js/sk_key_utils.js:
```diff
@@ -0,0 +1,367 @@
+              window.MozMutationObserver;
+      if (!MutationObserver) {
+        throw Error('MutationObserver not supported');
+      }
```
Remember that you're not writing code where cross-browser compatibility is going to matter. This is only going to run in gecko 37 and above.  You can just use MutationObserver without any of the stuff above.

pgmanutd replied:
Right, Removed.

davidflanagan commented on shared/js/sk_key_utils.js:
```diff
@@ -0,0 +1,367 @@
+        // TODO: Do we need to initialize again or handle each change.
+        //       Handling each change might be difficult
+        // init();
+
```
You shouldn't have to re-initialize; that would be too slow.  Think about the kinds of changes we care about:

1) if the app changes the textContent of any of the buttons with accesskey on them, then we need to mirror that change in the soft key labels.  You can do that by creating one MutationObserver for each of those buttons (never more than 3 of them).  Those observers only have to pay attention to characterdata changes, I think.

2) You also need to know if accesskey attributes are added or removed anywhere in the document. And if that happens, you'll need to change the appropriate soft key.

3) You probably also have to monitor changes to the document itself, to see whether the elements with accesskey set are removed from the document.  Or maybe you can just monitor the 1, 2, or 3 current elements directly (observing the child list of the parent, maybe?)

4) You need to know if a new element with an access key attribute is added. You may not need to do anything for this: it could be that you'll catch those things just by monitoring attribute changes.

You'll want to be sure to not start monitoring until after the DOMContentLoaded event or after the load event. Otherwise your observers will be triggered all the time as the document is being built. If you use <script deferred> then that will probably be fine

pgmanutd replied:
Did some changes. You can review.
Handling 4 scenarios:
- textcontent changed
- element with an accesskey added // still pending
- element with an accesskey removed
- accesskey changed

davidflanagan commented on shared/js/sk_key_utils.js:
```diff
@@ -0,0 +1,367 @@
+(function (global) {
+  'use strict';
+
+  var softKeyModule = (function () {
```
This extra layer of nesting doesn't seem necessary.

pgmanutd replied:
Right, Removed

davidflanagan commented on shared/js/sk_key_utils.js:
```diff
@@ -0,0 +1,367 @@
+    };
+
+    // Storing the reference to softkeys
+    var sk_keys = {
```
This object stores references to the DOM elements that the soft keys reference.  A name like sk_elements might be clearer than sk_keys.

And since this is all wrapped up inside a private function, there isn't any need for sk_ prefixes here. The code might be simpler to read without them.

pgmanutd replied:
Right, Changed.

davidflanagan commented on shared/js/sk_key_utils.js:
```diff
@@ -0,0 +1,367 @@
+      cancelKey: SK_KEYS_MAPPING.LSK,
+      cancelKeyText: 'Cancel',
+      cancelKeyl10nId: 'cancel'
+    };
```
I'm not going to review the option menu stuff just now. Let's get the basic soft key feature working first, then we'll figure out the context menu stuff.

pgmanutd replied:
Ok, but that part is working fine. You can review if you want.

davidflanagan commented on shared/js/sk_key_utils.js:
```diff
@@ -0,0 +1,367 @@
+
+    return {
+      init: init
+    };
```
If you don't bother with the second layer of wrapping then you don't need to return this, but just call init() down below.

pgmanutd replied:
Right, Done.

davidflanagan commented on shared/js/sk_key_utils.js:
```diff
@@ -0,0 +1,367 @@
+    function init() {
+
+      // Calling destroy to destroy previously initialized variables
+      destroy();
```
I hope that this destroy() method is never actually needed. I'd hope that it would be initialized once, and then modified as needed after that.  

It can be useful to have init and destroy methods for unit tests, though..

pgmanutd replied:
Yes, it is once. Added for unit tests.

davidflanagan commented on shared/js/sk_key_utils.js:
```diff
@@ -0,0 +1,367 @@
+      destroy();
+
+      // Selecting the softkey DOM element
+      var softKey = document.querySelector('[role="softkey"]');
```
I don't like this part.  The idea here is that we want to take an existing app, which might have a back button in the upper left and some other button somewhere completely unrelated, and make those buttons work with soft keys. We don't want the app author to have to place their buttons in a special div to make them work with this softkey module.

pgmanutd replied:
Right, changed accordingly.

davidflanagan commented on shared/js/sk_key_utils.js:
```diff
@@ -0,0 +1,367 @@
+
+      // Container will be created to handle options
+      // menu and soft keys
+      createSoftKeyContainer();
```
It would probably be helpful to think of the softkey container as a separate module (or even a Web Component) that is responsible for displaying the soft key labels and handling the key events for the soft keys

pgmanutd replied:
Will think on that.

davidflanagan commented on shared/js/sk_key_utils.js:
```diff
@@ -0,0 +1,367 @@
+
+            // Softkeys will be created for NON-Touch UI only
+            // A #IFDEF check is required
+            createSoftKeyButtons(sk_keys[k], k);
```
If this function creates a single button, it should use the singular, not the plural.

davidflanagan replied:
And after creating a softkey for the element the other thing we need to do is hide the element. The element expects a click, but we don't have any way to click it, so we're replacing it with the soft key. This means that the unclickable element should not appear to the user.

It might be simplest, though, to just leave this to CSS and include styles that hide elements with accesskey attributes. That would give apps the ability to customize things. Some would be hidden and others would be display:none

pgmanutd replied:
Hiding the element in "createSoftKeyButton" function by adding a class "sk-hide-button"

davidflanagan commented on shared/js/sk_key_utils.js:
```diff
@@ -0,0 +1,367 @@
+          if (sk_keys[k]) {
+
+            // Softkeys will be created for NON-Touch UI only
+            // A #IFDEF check is required
```
We'll have to figure that out somehow. But let's just focus on non-touch devices for now.

pgmanutd replied:
Yes, but this part is working fine.

davidflanagan commented on shared/js/sk_key_utils.js:
```diff
@@ -0,0 +1,367 @@
+
+            // Adding Event Listener
+            sk_keys[k].addEventListener('click',
+                    handleSoftKey,
```
This part is completely wrong. Remember that these elements with accesskey attributes are existing parts of a working app. They already have their own click handlers defined.

When the user presses a softkey, we will receive a keydown event. When that happens we want to trigger the existing event handlers on those elements.

pgmanutd replied:
Yes, handling the click listeners for context menu only and triggering the clicks on other buttons.

davidflanagan commented on shared/js/sk_key_utils.js:
```diff
@@ -0,0 +1,367 @@
+            // Adding Event Listener
+            sk_keys[k].addEventListener('click',
+                    handleSoftKey,
+                    false);
```
Just FYI, this third argument used to be required, but it has been legal to omit it for a couple of years now.  So you only need it if you're passing true. You can leave off the false.

pgmanutd replied:
Removed

davidflanagan commented on shared/js/sk_key_utils.js:
```diff
@@ -0,0 +1,367 @@
+
+          // Selecting the button using accesskey
+          sk_keys[k] =
+                  softKey.querySelector('[accesskey="' + SK_KEYS_MAPPING[k] + '"]');
```
Most of our Gaia apps are single-page apps where multiple views are defined in the document, but only one is visible at a time.  So a full implementation of this soft key system may need to use querySelectorAll() to find all elements with this accesskey and then check to see which one is visible  (or which one is in a container that is visible).

davidflanagan replied:
That would also mean that we'd have to track visibility changes.  I'm not sure if we can do that with MutationObserver or not...

pgmanutd replied:
Initially, only 1 view will be visible. So queryselector will be fine.

davidflanagan commented on shared/js/sk_key_utils.js:
```diff
@@ -0,0 +1,367 @@
+    const SK_KEYS_MAPPING = {
+      LSK: 'a',
+      CSK: 's',
+      RSK: 'd'
```
Personally, I'd use F1 for LSK, Enter for CSK and F2 for RSK, but that doesn't matter for prototyping purposes.

pgmanutd replied:
Changed accordingly

davidflanagan commented on shared/js/sk_key_utils.js:
```diff
@@ -0,0 +1,367 @@
+     * @return void 0
+     */
+    function createSoftKeyContainer() {
+      var sk_button_container = null;
```
I'd make this a top-level variable like sk_container so that you don't have to querySelector to find it.

pgmanutd replied:
Done.

davidflanagan commented on shared/js/sk_key_utils.js:
```diff
@@ -0,0 +1,367 @@
+      sk_button_container.classList.add('sk-button-container');
+      sk_container.appendChild(sk_button_container);
+
+      document.body.appendChild(sk_container);
```
I think it would be simplest to just go ahead and create the three buttons here as well, rather than creating them dynamically.

pgmanutd replied:
Need to read textcontent from the buttons and hide them, that's why dynamically creating softkeys.

davidflanagan commented on shared/js/sk_key_utils.js:
```diff
@@ -0,0 +1,367 @@
+      button.classList.add('sk-buttons', keyType);
+      sk_container
+              .querySelector('.sk-button-container')
+              .appendChild(button);
```
Note that you never remove these buttons once they are added. I'd create all 3 buttons when you create the container. Then, as the mutation observer detects changes you just have to change the text of the buttons.

pgmanutd replied:
Need to read textcontent from the buttons and hide them, that's why dynamically creating softkeys.

davidflanagan commented on shared/js/sk_key_utils.js:
```diff
@@ -0,0 +1,367 @@
+              .appendChild(button);
+
+      // Hide the buttons for Non-Touch UI
+      key.classList.add('sk-hide-button');
```
If we're in a non-touch UI, then we should not create the buttons in the first place!  And if you want the option to hide them via CSS, it is the container that you'd want to hide, not the buttons themselves.

pgmanutd replied:
// Hide the buttons for Non-Touch UI
      key.classList.add('sk-hide-button');

This part is hiding the original buttons and replacing with softkeys

davidflanagan commented on shared/js/sk_key_utils.js:
```diff
@@ -0,0 +1,367 @@
+     * @param N/A
+     * @return void 0
+     */
+    function handleSoftKey() {
```
I don't want to think about contextmenus yet in this review, so I'm ignoring most of this function.

You need event handlers to handle keydown events for the soft keys, and when those occur you probably can just call .click() on the elements that the softkeys map to.

pgmanutd replied:
Done.

davidflanagan commented on shared/js/sk_key_utils.js:
```diff
@@ -0,0 +1,367 @@
+     * @param {object} options: Pass options
+     * @return void 0
+     */
+    function showOptionsMenu(options) {
```
I'm ignoring this bit for now. It might be worth having a separate module for displaying menus, if that is possible.

davidflanagan commented on shared/js/sk_key_utils.js:
```diff
@@ -0,0 +1,367 @@
+      sk_container = document.createElement('div');
+      sk_container.dataset.subtype = 'menu-parent';
+      sk_container.tabIndex = -1;
+      sk_container.setAttribute('role', 'dialog');
```
I don't understand why we'd want role="dialog" here

pgmanutd replied:
Removed. Copy pasted the code. Forgot to remove.

davidflanagan commented on shared/style/sk_key_utils.css:
```diff
@@ -0,0 +1,379 @@
+
+form[role="dialog"][data-subtype="menu"].visible {
+  visibility: visible;
+}
```
Using generic selectors like this risks breaking other parts of the UI. I'd just give the soft key container element an explicit id or class and using that here.

pgmanutd replied:
Done. Added parent class.

davidflanagan commented on shared/style/sk_key_utils.css:
```diff
@@ -0,0 +1,379 @@
+/* Action Menu 
+ * Copied from shared/style/action_menu.css 
+ * Some cleanup needs to be done
+*/
```
This would all probably be better in a separate CSS file, I think.  Ideally just document that this module requires shared/style/action_menu.css.  Or customize it and create a simplified context_menu.css file that has only what we need here.

Commit:
Made the changes acc. to David's Comments

Added Mutation observer section.
Missing: Need to handle when an element
added to the DOM with an accesskey.

pgmanutd commented:
@davidflanagan : Added a new commit. Please review.

Added Mutation observer section.
Missing: Need to handle when an element
added to the DOM with an accesskey.

SmirnovAndrey commented:
I tried to use current version of soft keys implementation. We have two screens in application. And softkeys must be displayed only on one of the screens.
Can you explain me how I can do this, if current functionality provides it? And I don't find functionality that provides to add functions/callbacks on buttons.
Also I see that this module has only methods. It would be better if it will be an object what we can expand in every application for our needs and use it's basic functions, for example for handling callbacks, showing softkeys, changing texts on buttons, destroying, getting information about softkeys.

pgmanutd commented:
@SmirnovAndrey :
Earlier we implemented the same way you mentioned, but David rejected it.

Check this bug:-
https://bugzilla.mozilla.org/show_bug.cgi?id=1142855

Now your you should have something like this in your HTML file (per screen)

&lt;button type="button" accesskey="a"&gt;Back&lt;/button&gt;
&lt;button type="button" accesskey="s"&gt;OK&lt;/button&gt;

&lt;button type="button" accesskey="d" contextmenu="menu1"&gt;Options&lt;/button&gt;

&lt;menu id="menu1" header="Header Text"&gt;

  &lt;menuitem data-menuid="option1" label="Option1"
    icon="img/icons/icon48x48.png"&gt;&lt;/menuitem&gt;
  &lt;menuitem data-menuid="option2" label="Option2"
    icon="img/icons/icon48x48.png"&gt;&lt;/menuitem&gt;
  &lt;menuitem data-menuid="option3" label="Option3"
    icon="img/icons/icon48x48.png"&gt;&lt;/menuitem&gt;
  &lt;menuitem data-menuid="option4" label="Option4"
    icon="img/icons/icon48x48.png"&gt;&lt;/menuitem&gt;

&lt;/menu&gt;
- First three are softkey buttons with an accesskey which will trigger the click event on that button. You need to attach your own click eventlistener for handling click events.
- If you want to display a Menu, add a contextmenu attribute pointing to that menu element.

This is not the final commit, and still need some changes (waiting for feedback from David).

Commit:
Bug 1142855 - Minor bug fixes

Changed zindex, and fixed 2 bugs related to js file

ruvsmirnov commented:
@pgmanutd , 
- we need to show softkeys only for one screen in HomeScreen app and hide them for others. I found the only way how to implement this functionality - to destroy softkey control on one screen and init it on another screen again, but it doesn't seem to be a good approach. May be a better approach would be to add show and hide methods that will provide this functionality by passing show/hide class or some attribute that shows and hides softkeys. What do you say?
- Is there any way to simply rename buttons without destroying the control? The only way that I found is to destroy softkeys and init softkeys again with new name. Probably it would be more useful to add some function, that would do this without reinitialization of the control.

sergkashin commented on shared/js/sk_key_utils.js:
```diff
@@ -0,0 +1,419 @@
+            break;
+          }
+        }
+        else {
```
else should stay on the same line (line 162) as `}` bracket.

AlekoNN commented:
Closed since a new PR#31 submitted.
